downstairs/src/stats.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2021 Oxide Computer Company
+use super::*;
+
+use dropshot::{ConfigDropshot, ConfigLogging, ConfigLoggingLevel};
+use omicron_common::api::internal::nexus::ProducerEndpoint;
+use oximeter::{
+    types::{Cumulative, Sample},
+    Error, Metric, Producer, Target,
+};
+use oximeter_producer::{Config, Server};
+
+// These structs are used to construct the required stats for Oximeter.
+#[derive(Debug, Copy, Clone, Target)]
+pub struct DsStat {
+    // The UUID of the downstairs
+    pub ds_uuid: Uuid,
+}
+#[derive(Debug, Copy, Clone, Metric)]
+pub struct DsConnect {
+    // Count of times this downstairs has started a connection to an upstairs
+    #[datum]
+    pub count: Cumulative<i64>,
+}
+#[derive(Debug, Copy, Clone, Metric)]
+pub struct DsWrite {
+    // Count of region writes this downstairs has completed
+    #[datum]
+    pub count: Cumulative<i64>,
+}
+#[derive(Debug, Copy, Clone, Metric)]
+pub struct DsRead {
+    // Count of region reads this downstairs has completed
+    #[datum]
+    pub count: Cumulative<i64>,
+}
+#[derive(Debug, Copy, Clone, Metric)]
+pub struct DsFlush {
+    // Count of region flushes this downstairs has completed
+    #[datum]
+    pub count: Cumulative<i64>,
+}
+
+// All the counter stats in one struct.
+#[derive(Clone, Debug)]
+pub struct DsCountStat {
+    stat_name: DsStat,
+    up_connect_count: Vec<DsConnect>,
+    write_count: Vec<DsWrite>,
+    read_count: Vec<DsRead>,
+    flush_count: Vec<DsFlush>,
+}
+
+impl DsCountStat {
+    pub fn new(ds_uuid: Uuid) -> Self {
+        DsCountStat {
+            stat_name: DsStat { ds_uuid },
+            up_connect_count: vec![DsConnect {
```

Comment:
Since `Cumulative` implements `Default`, you could add a derive for `Default` to `DsConnect` and friends. Then this would be:

```
DsCountStat {
    stat_name: DsStat { ds_uuid },
    ..default::Default(),
}
```

Reply:
Nevermind, these are `Vec`s, which would default to empty, I believe.

Reply:
Actually, why are these vectors? It doesn't seem like there's ever more than one element, but maybe there will be in the future?

Reply:
I made them `Vec` because I thought I had to :)
I had some trouble when I tried to make them a `Cumulative`.  Now that everything else is working, I'll see if I can get them back as a single element

Reply:
Cool, making them just `Cumulative` "should" work. It's also probably possible to restructure some of this to avoid the explicit lock as well, since the `Cumulative` is thread-safe. But that's probably not worth it at this point :)

Reply:
I've updated these all to be just a single `Cumulative`